Reorganized layout section and properties

mattgarrish opened the pull request:
This pull request is a continuation of #2842 but rebased on the current draft to avoid merge conflicts.

A brief recap from #2842:

> - moved the layout section up ahead of content documents
> - added a layout types section to explain each of three main layout types
> - moved all the property definition text back into appendix D.5
> - kept synthetic spreads as a separate subsection
> - moved the section on fixed layout documents and setting their dimensions into the content documents section.
> - dropped the "roll document"

New to this pull request:

- I've integrated the changes to remove rendition:orientation and its spine overrides from #2831 (adding it as obsolete but conforming and leaving rendition:viewport as deprecated)
- I've added a layout section to the reading systems specification and put the requirements I could find for each layout type together
- I added a couple of requirements for rolls based off the existing authoring spec prose and our discussions about no layout control properties applying to it

A bit more controversially, perhaps, I've noted that images in spine are a type of fixed-layout document even if they require fallbacks, and added additional caveats about vendor acceptance and accessibility in the existing note about fxl support. I also added a section to the authoring spec that the icb dimensions will be obtained from the image metadata and to refer to the image format definition for how that's done. I similarly stated for reading systems that if they opt to support image formats in the spine that's where the dimensions are obtained. That should be all we need to say so that we don't have to classify images in spine as separate from fxl documents for either pre-paginated or roll layouts.

Still to do before we look to merge is to resolve on the other properties and overrides we're deprecating so I can take them out without worrying that they'll have to be re-added. This will not only affect the vocabulary, but could mean making spread placement specific to pre-paginated layouts if we decide they aren't useful for reflowable documents.

Fixes #2841 
Fixes #2754 
Fixes #2753 
Fixes #2751 
Fixes #2750 
Fixes #2749 
Fixes #2847 

See:

* For EPUB 3.4:
    * [Preview](https://raw.githack.com/w3c/epub-specs/fix/reorg-layouts/epub34/authoring/index.html)
    * [Diff](https://services.w3.org/htmldiff?doc1=https://labs.w3.org/spec-generator/%3Ftype=respec%26url=https://w3c.github.io/epub-specs/epub34/authoring/index.html&doc2=https://labs.w3.org/spec-generator/%3Ftype=respec%26url=https://raw.githack.com/w3c/epub-specs/fix/reorg-layouts/epub34/authoring/index.html)
* For EPUB 3.4 Reading Systems:
    * [Preview](https://raw.githack.com/w3c/epub-specs/fix/reorg-layouts/epub34/rs/index.html)
    * [Diff](https://services.w3.org/htmldiff?doc1=https://labs.w3.org/spec-generator/%3Ftype=respec%26url=https://w3c.github.io/epub-specs/epub34/rs/index.html&doc2=https://labs.w3.org/spec-generator/%3Ftype=respec%26url=https://raw.githack.com/w3c/epub-specs/fix/reorg-layouts/epub34/rs/index.html)

(Added preview links for RS, too, IH)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/epub-specs/pull/2844.html" title="Last updated on Dec 18, 2025, 2:14 PM UTC (2445a58)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/epub-specs/2844/ba46f10...2445a58.html" title="Last updated on Dec 18, 2025, 2:14 PM UTC (2445a58)">Diff</a>

mattgarrish committed:
minor cleanup

mattgarrish committed:
add unsaved changes

mattgarrish committed:
fix duplicate page-spread-* id errors in vocabulary

HadrienGardeur commented:
> Don't we have spread placement for reflowable to handle the case of mixing pre-paginated documents or having "reflowable" SVGs in the spine? (what is a reflowable svg?)

If you mix pre-paginated and reflow together, they won't be displayed side by side in a spread since the concept of synthetic spread is completely foreign to reflowable documents.

I think that there's confusion about two separate concepts:
- when a reflowable document is paginated, this is usually done with columns where the RS treats each reflowable resource as a long horizontal roll where we move the viewport to view one or two columns at a time
- whereas a synthetic spread is usually handled using two webviews or iframes side by side in a container

mattgarrish commented:
> If you mix pre-paginated and reflow together, they won't be displayed side by side in a spread since the concept of synthetic spread is completely foreign to reflowable documents.

Ya, that's true. We do say the properties are only for synthetic spread placement, not simulated two-page reading. But we still have to allow for synthetic spread placement in reflowable layouts because of overrides to contain the documents in the spread. 

So, yes, spreads may only be created for pre-paginated documents, but they aren't exclusive to a global pre-paginated layout. We can, of course, define them under pre-paginated layouts and then have the reflowable refer back to the section for embedded spreads.

It just seems like a missed opportunity that you couldn't have a fxl document in one slot and a reflowable in another, but perhaps there's no practical use for it, and it undoubtedly would complicate rendering. But think of an art book, for example, where you could have a painting continuously in view in one slot with a scrolling reflowable document describing it in the other.

iherman commented on epub34/authoring/index.html:
```diff
@@ -5829,6 +5828,410 @@ <h4>The <code>itemref</code> element</h4>
+					</ul>
+
+					<p>All other properties in the Package rendering vocabulary are ignored for reflowable spine
+						items.</p>
```
```suggestion
					<p>All other properties in the Package rendering vocabulary are ignored for reflowable spine
						items.</p>
```

Shouldn't we use a more normative language for this type of statement (it is repeated in the other layout subsection). Something like "Other properties in the Package rendering vocabulary _MUST NOT_ be used in spine items". If we are afraid of making some valid documents suddenly invalid, we could used _SHOULD NOT_.

mattgarrish replied:
Those lists are a placeholder for now until we figure out what properties are even surviving this revision. I had all the other properties listed but then took them out to match what looks like will be left after we cull the vocabulary.

But going back to when the list had global properties in it, not just the remaining spine overrides, the problem is that you can mix layouts using spine overrides. So say I want to globally set a value for all the pre-paginated overrides I've added to my globally reflowable layout. A restriction that you must not or should not set pre-paginated properties in reflowable layouts would make it invalid and you'd have to repeat spine overrides everywhere. They're fine to set but they're ignored for reflowable documents by reading systems... hence the statement.

There's also the problem that warnings are as bad as errors for vendor ingestion, so it's probably better to leave the requirement to ignore other properties on reading systems than restrict their presence.

sueneu commented:
> For someone not well versed in FXL (sorry, pre-paginated or roll), the section on spreads is really difficult to read. I wonder whether it is possible to reuse, or at least refer to, some of the diagrams we created in 3.3. I realize they are present in the vocabulary definition, but they should, somehow, be present in this section as well. Or, if necessary, create some versions of the diagrams.
> 
> (Yes, this could be done in a later PR, once this one is finalized.)

Readability is a worthwhile goal but maybe too many edits for this PR. Would it be worth making a new issue about this so we can bring it up after this PR is merged?

HadrienGardeur committed:
Removed the concept of "true spreads"

This is based on a resolution from the TPAC meeting: https://w3c.github.io/pm-wg/minutes/2025-11-10-f2f.html#1a21:~:text=RESOLUTION%3A%20Remove%20text%20saying%20that%20presence%20of%20spread%20information%20implies%20that%20the%20pages%20SHOULD%20be%20laid%20out%20as%20a%20spread%2C%20clarify%20that%20either%20presentation%20equally%20acceptable.

mattgarrish committed:
remove the properties expected to be deprecated;
remove for now the sentence that says spread placement applies to reflowable;
move the spread placement section under pre-paginated layouts and remove unneeded layout types section;
various cleanup on remaining property tables and references to removed properties

mattgarrish committed:
Merge branch 'fix/reorg-layouts' of https://github.com/w3c/epub-specs into fix/reorg-layouts

# Conflicts:
#	epub34/authoring/index.html

mattgarrish committed:
remove the flow and spread property requirements from the reading system spec

HadrienGardeur commented:
@mattgarrish I tweaked some of the language related to spread placement in a previous commit but I can see that you're still adjusting that section so I'll comment instead.

> When a [reading system](https://pr-preview.s3.amazonaws.com/w3c/epub-specs/pull/2844.html#dfn-epub-reading-system) renders a [synthetic spread](https://pr-preview.s3.amazonaws.com/w3c/epub-specs/pull/2844.html#dfn-synthetic-spread), the default behavior is to populate the spread by rendering the next [EPUB content document](https://pr-preview.s3.amazonaws.com/w3c/epub-specs/pull/2844.html#dfn-epub-content-document) in the next available unpopulated [viewport](https://pr-preview.s3.amazonaws.com/w3c/epub-specs/pull/2844.html#dfn-viewport), where the next available viewport is determined by the given [page progression direction](https://pr-preview.s3.amazonaws.com/w3c/epub-specs/pull/2844.html#attrdef-spine-page-progression-direction) or by local declarations within [EPUB content documents](https://pr-preview.s3.amazonaws.com/w3c/epub-specs/pull/2844.html#dfn-epub-content-document).

This might be a good place to also indicate that without spread placement information, there's an expectation that the first resource should be displayed on its own.

We have [a resolution](https://w3c.github.io/pm-wg/minutes/2025-11-10-f2f.html#1a21:~:text=RESOLUTION%3A%20Adapt%20spec%20language%20in%20Core%20and%20Reading%20Systems%20to%20say%20that%20if%20no%20information%20about%20placement%20in%20spreads%2C%20the%20first%20resource%20should%20be%20a%20page%20on%20its%20own.) on day 1 of TPAC specifically about that:

> RESOLUTION: Adapt spec language in Core and Reading Systems to say that if no information about placement in spreads, the first resource should be a page on its own.

This was also an action item for me to open a PR to cover this change, let me know if you'd prefer that.

> Although it is common practice to specify the use of a spread in certain device orientations, the content itself does not represent a true spread — two consecutive pages that have to be rendered side-by-side for readability, such as a two-page map.

Since we're making `rendition:spread` obsolete, talking about a "common practice to specify the use of a spread in certain device orientations" feels weird.

> To indicate that two consecutive pages represent a true spread, the rendition:page-spread-left and rendition:page-spread-right properties SHOULD be set on the [spine](https://pr-preview.s3.amazonaws.com/w3c/epub-specs/pull/2844.html#dfn-epub-spine) items for the two adjacent EPUB content documents, and the properties omitted on spine items where one-up or two-up presentation is equally acceptable.

We've agreed to drop that concept from the specification at TPAC, there was [a resolution](https://w3c.github.io/pm-wg/minutes/2025-11-10-f2f.html#1a21:~:text=RESOLUTION%3A%20Remove%20text%20saying%20that%20presence%20of%20spread%20information%20implies%20that%20the%20pages%20SHOULD%20be%20laid%20out%20as%20a%20spread%2C%20clarify%20that%20either%20presentation%20equally%20acceptable.) for this on day 1:

> RESOLUTION: Remove text saying that presence of spread information implies that the pages SHOULD be laid out as a spread, clarify that either presentation equally acceptable.

mattgarrish committed:
add section for reading system handling of obsolete but conforming features

mattgarrish committed:
additional cleanup per Hadrien's comment and to the RS spec - including adding a recommendation for rendition:page-spread-center handling

mattgarrish commented:
I've made some additional fixes now, including removing that paragraph. I did some rewriting to the first few paragraphs of the section, too. The reading system spec still had some mixing of requirements for reflowable documents in spreads, and I noticed we said nothing at all about how to handle rendition:page-spread-center, so I tried to clean it up as best I could, as well. Again, feel free to directly edit anything if you think it can stand for improvement.

mattgarrish committed:
adds "fixed layout" grouping section;
switches "obsolete but conforming" to "outdated";
fixes a couple of broken respec shorthand references

mattgarrish committed:
additional clarifications to authoring fxl description;
add "fixed layout" parent structure to rs spec;
simplify rs support requirement for fxl docs

sueneu reviewed (commented):
“outdated” is much simpler and works well here. Is there a place where our use of it is defined for people who might want that? 

Maybe in Authoring Shorthands?
`obsolete: features or attributes that are no longer in common use but are still considered conforming for backwards compatibility`

We should probably make it clear in the RS spec that obsolete features should still be supported.

mattgarrish commented:
> “outdated” is much simpler and works well here. Is there a place where our use of it is defined for people who might want that?

Yes, we have an appendix that explains the meanings and validation expectations for both outdated and deprecated. If you're reading the github diff and see the syntax `[= ... =]` around words that's a respec shorthand for linking terms to their definitions, so every instance of deprecated and outdated will get turned into hyperlinks in the rendered document.


> We should probably make it clear in the RS spec that obsolete features should still be supported.

I'm updating the reading system appendix in this PR, as well. We used to only say that they may support deprecated features, but I've added separate subsections for each category. The outdated features still have that subset of EPUB 2 legacy features that reading systems must not support that needs to be called out.

HadrienGardeur commented:
> But it is much more convenient to have to set it at one place.

… and much more dangerous as well. I've seen many publications that set this property to `none` for no good reason, which negatively impacts the UX.

> Where is it specified in the specification that, for images in the spine, the spread should be none by default. Because if it is not specified, then it is not, strictly speaking, a bug of Thorium...

That's not the problem IMO.

Thorium Desktop most likely defaults to "auto" but I suspect a bug when laying out spreads with images in spine.

mattgarrish commented:
> I believe our best way forward is to merge this asap

I'm getting down to minor cleanup stuff, but I still want to make sure everything is reasonably consistent. I'm hopping between documents today but I'll definitely flip this out of draft status before the end of the weekend and see if there's any final feedback.

mattgarrish committed:
final cleanup

mattgarrish committed:
updated scripting layout requirements

mattgarrish committed:
minor cleanup of spread mixing requirements

mattgarrish committed:
adding roll example image

iherman commented:
> 
> I'd suggest opening a new issue for this, though, as we should probably discuss what is realistic and/or if keeping the requirements as aspirational is worth it.

+1

mattgarrish committed:
minor wording tweak

mattgarrish committed:
unrelated fix to the pageBreakSource section title to standardize with the rest of the meta properties

mattgarrish committed:
additional edits for initial spread placement

mattgarrish committed:
minor clarification

mattgarrish committed:
minor edit

mattgarrish committed:
rewrite author layout intentions to not use normative keywords

mattgarrish committed:
rewrite initial page placement to use page-spread-center

bduga commented:
> In this case, the reading system will render the first document on its own (equivalent to setting the rendition:page-spread-center property).

Hang on - I don't agree with that, and I don't think it is what we decided on. I consider the exact placement to be a stylistic one, and there is no practical reason why it must be centered. So while this is bad: [1,2],[2,3]... there is no practical difference between [1], [2,3]... and [-,1],[2,3]... In fact, Play Books Web uses the former while PB Android uses the latter. I don't think we should dictate a specific styling, we just need to make sure there spreads are aligned with the correct pages.

iherman commented:
Absolutely super minor change before merge: in the authoring change section the 10 October entry refers to the classification of obsolete but non conforming; we now have to note the change to "outdated". I think a separate entry should also be added referring to the reorganization, in particular with the layout classification "reflowable-fixed layout" and fxl classification of "prepaginated-roll". Maybe the fact that spread related properties are now very stricly bound to pre-paginated could also be noted.

(Note to myself: it will be time to refresh the overview document!)

iherman commented:
Another minor point (it could be added as a separate issue, or reopening https://github.com/w3c/epub-specs/issues/2792): while we agreed on TPAC about that no normative changes will be done regarding images in the spine, I believe we also said that we would add a note somewhere in the document about the fact that it is possible to create an FXL document (pre-paginated or roll) with images in the spine, provided there is a proper fallback for each image.

@reidmore-online @HadrienGardeur

mattgarrish committed:
add additional change log entries;
delete align-x-center property def (unsaved from last commit)

mattgarrish committed:
add change log entry for obs but conforming name change

mattgarrish committed:
rename "unsupported features" in rs spec to match the "obsolete features" name we use in the authoring spec

HadrienGardeur committed:
Expanded example for roll layout

Added manifest and spine, with images in spine that fallback to HTML

mattgarrish committed:
minor wording consistency fix

mattgarrish committed:
Merge branch 'fix/reorg-layouts' of https://github.com/w3c/epub-specs into fix/reorg-layouts

mattgarrish committed:
add note about pre-paginated + scrolled-continuous;
a few other minor tweaks

mattgarrish commented:
> * and we're still missing a note that mentions what some Japanese publishers are currently using for roll layout

I've added the following in the last commit:

> Prior to the introduction of roll layouts, some Japanese publishers combined a [pre-paginated layout](http://localhost:120/epub34/rs/#sec-pre-paginated-layouts) with a [rendition:flow property](https://www.w3.org/TR/epub-33/#flow) [[epub-33](http://localhost:120/epub34/rs/#bib-epub-33)] set to the value "scrolled-continuous" to create rolls. If reading systems developers intend to support this content, they will need to treat the combination as synonymous with a roll layout being declared.

Feel free to adjust it, though.

rdeltour commented on epub34/authoring/index.html:
```diff
@@ -5829,6 +5827,429 @@ <h4>The <code>itemref</code> element</h4>
+
+					<p>To declare that an EPUB publication uses a pre-paginated fixed layout, the <a href="#layout"
+								><code>rendition:layout</code> property</a> MUST be declared in a <code>meta</code>
+						element that specifies the value <dfn id="def-layout-pre-paginated">pre-paginated</dfn>.</p>
```
I personally like Matt's proposed rephrasing better than the must.

To me this is about what is a requirement vs what is a definition. In fact, the presence of the layout properties in the package document is what **defines** a document as fixed-layout or not. 

With a MUST statement, my mindset goes "when I have a fixed-layout document, then I MUST declared the property. But wait, why is this a fixed layout document? Oh, because I declared the property." And then it is looping. So I better understand this as: "a content document is fixed-layout when it is declared with this property"; it is not a requirement, it is its definition.

That's said, it is mostly editorial and spec purity, so I definitely won't die on that hill 😊

rdeltour commented:
Just a nit to pick but I'm not a big fan of the introduction of the "outdated" terminology.
- Before we had : "obsolete" = "obsolete but conforming" + "deprecated"
- now we have "obsolete" = "outdated" + "deprecated"

I'm not convinced it is much clearer without looking into the details of how these terms are defined. Or maybe it is obvious to English native speakers?

I would find it more intelligible as:
- "obsolete" = "obsolete but conforming" + "obsolete and not recommended".
(or the equivalent with "outdated" or "deprecated" instead of "obsolete", I don't have a strong preference here).

i.e. only have one keyword, with the two variants refined by qualifiers.

mattgarrish committed:
move align-x-center to deprecated properties
